cdc/log.cc

```diff
@@ -479,6 +479,134 @@ api::timestamp_type find_timestamp(const schema& s, const mutation& m) {
     throw std::runtime_error("cdc: could not find timestamp of mutation");
 }
 
+// iterators for collection merge
+template<typename T>
+class collection_iterator : public std::iterator<std::input_iterator_tag, const T> {
+    bytes_view _v, _next;
+    size_t _rem = 0;
+    T _current;
+public:
+    collection_iterator(bytes_view_opt v = {}) 
+        : _v(v.value_or(bytes_view{}))
+        , _rem(_v.empty() ? 0 : read_collection_size(_v, cql_serialization_format::internal()))
+    {
+        if (_rem != 0) {
+            parse();
+        }
+    }
+    collection_iterator(const collection_iterator&) = default;
+    const T& operator*() const {
+        return _current;
+    }
+    const T* operator->() const {
+        return &_current;
+    }
+    collection_iterator& operator++() {
+        next();
+        if (_rem != 0) {
+            parse();
+        } else {
+            _current = {};
+        }
+        return *this;
+    }
+    collection_iterator operator++(int) {
+        auto v = *this;
+        ++(*this);
+        return v;
+    }
+    bool operator==(const collection_iterator& x) const {
+        return _v == x._v;
+    }
+    bool operator!=(const collection_iterator& x) const {
+        return !(*this == x);
+    }
+private:
+    void next() {
+        --_rem;
+        _v = _next;
+    }
+    void parse();
+};
+
+template<>
+void collection_iterator<std::pair<bytes_view, bytes_view>>::parse() {
+    assert(_rem > 0);
+    _next = _v;
+    auto k = read_collection_value(_next, cql_serialization_format::internal());
+    auto v = read_collection_value(_next, cql_serialization_format::internal());
+    _current = std::make_pair(k, v);
+}
+
+template<>
+void collection_iterator<bytes_view>::parse() {
+    assert(_rem > 0);
+    _next = _v;
+    auto k = read_collection_value(_next, cql_serialization_format::internal());
+    _current = k;
+}
+
+template<typename Container, typename T>
+class maybe_back_insert_iterator : public std::back_insert_iterator<Container> {
+    const abstract_type& _type;
+    collection_iterator<T> _s, _e;
+public:
+    using value_type = typename Container::value_type;
+    maybe_back_insert_iterator(Container& c, const abstract_type& type, collection_iterator<T> s)
+        : std::back_insert_iterator<Container>(c)
+        , _type(type)
+        , _s(s)
+    {}
+    maybe_back_insert_iterator& operator*() {
+        return *this;
+    }
+    maybe_back_insert_iterator& operator=(const value_type& v) {
+        if (!find(v)) {
+            std::back_insert_iterator<Container>::operator=(v);
+        }
+        return *this;
+    }
+    maybe_back_insert_iterator& operator=(value_type&& v) {
+        if (!find(v)) {
+            std::back_insert_iterator<Container>::operator=(std::move(v));
+        }
+        return *this;
+    }
+private:
+    bool find(const value_type& v) {
+       // cheating, because we know we only append unique values (see below).
```

Comment:
I didn't understand what this "cheat" means, nor why find() modifies _s. Is this the cheat? And where below should I see?

Reply:
The cheat is modifying _s, yes. We know we only use the filter with `set_union` of unique (no repeats) sets, so it is safe to reduce the filter range like I do. If this was a generalized type though, it would not be. 
For generalizing, it should instead use a predicate operator callback or something, but I opted away that in this case, because that has higher overhead.